File: src/main/java/AoC2017_14.java
import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.graph.BFS;
import com.github.pareronia.aoc.knothash.KnotHash;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.stream.Stream;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2017_14 extends SolutionBase<CharGrid, Long, Integer> {

    private static final char ON = '1';

    private AoC2017_14(final boolean debug) {
        super(debug);
    }

    public static AoC2017_14 create() {
        return new AoC2017_14(false);
    }

    public static AoC2017_14 createDebug() {
        return new AoC2017_14(true);
    }

    @Override
    protected CharGrid parseInput(final List<String> inputs) {
        return CharGrid.from(
                IntStream.range(0, 128)
                        .mapToObj(i -> inputs.getFirst() + "-" + i)
                        .map(KnotHash::binString)
                        .toList());
    }

    private Set<Cell> findRegion(final CharGrid grid, final Cell start) {
        final Function<Cell, Stream<Cell>> adjacent =
                cell -> grid.getCapitalNeighbours(cell).filter(n -> grid.getValue(n) == ON);
        return BFS.floodFill(start, adjacent);
    }

    @Override
    public Long solvePart1(final CharGrid grid) {
        return grid.getAllEqualTo(ON).count();
    }

    @Override
    public Integer solvePart2(final CharGrid grid) {
        final List<Set<Cell>> regions = new ArrayList<>();
        grid.getAllEqualTo(ON)
                .filter(c -> regions.stream().noneMatch(r -> r.contains(c)))
                .map(c -> findRegion(grid, c))
                .forEach(regions::add);
        return regions.size();
    }

    @Samples({
        @Sample(method = "part1", input = "flqrgnkx", expected = "8108"),
        @Sample(method = "part2", input = "flqrgnkx", expected = "1242"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }
}

File: src/main/python/AoC2017_14.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 14
#

from collections import deque
from collections.abc import Iterator
from aoc import my_aocd
from aoc.knothash import KnotHash
from aoc.grid import IntGrid, Cell

import aocd

ON = 1


def _hashes(input_: str) -> Iterator[str]:
    for i in range(128):
        yield KnotHash.bin_string(f"{input_}-{i}")


def _find_region(grid: IntGrid, start: Cell) -> set[Cell]:
    seen = set[Cell]()
    q = deque[Cell]()
    q.append(start)
    while len(q) > 0:
        cell = q.popleft()
        for n in grid.get_capital_neighbours(cell):
            if grid.get_value(n) == ON and n not in seen:
                q.append(n)
                seen.add(n)
    return seen


def part_1(inputs: tuple[str, ...]) -> int:
    return sum(1 for h in _hashes(inputs[0]) for c in h if int(c) == ON)


def part_2(inputs: tuple[str, ...]) -> int:
    grid = IntGrid([[int(c) for c in h] for h in _hashes(inputs[0])])
    regions = list[set[Cell]]()
    ans = 0
    for cell in grid.get_all_equal_to(ON):
        if not any(cell in r for r in regions):
            regions.append(_find_region(grid, cell))
            ans += 1
    return ans


TEST = """flqrgnkx""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2017, 14)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST) == 8108  # type:ignore[arg-type]
    assert part_2(TEST) == 1242  # type:ignore[arg-type]

    inputs = my_aocd.get_input(2017, 14, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == "__main__":
    main()
